wolfcrypt/src/dh.c

```diff
@@ -81,14 +83,51 @@ static word32 DiscreteLogWorkFactor(word32 n)
         return (word32)(2.4 * XPOW((double)n, 1.0/3.0) *
                 XPOW(XLOG((double)n), 2.0/3.0) - 5);
 }
+#endif /* WOLFSSL_DH_CONST*/
+
+
+/* if not using fixed points use DiscreteLogWorkFactor function for unsual size
+   otherwise round up on size needed */
+#ifndef WOLFSSL_DH_CONST
+    #define WOLFSSL_DH_ROUND(x)
+#else
+    #define WOLFSSL_DH_ROUND(x) \
+        do {                    \
+            if (x % 128) {      \
+                x &= 0xffffff80;\
+                x += 128;       \
+            }                   \
+        }                       \
+        while (0)
+#endif
 
 
 static int GeneratePrivate(DhKey* key, WC_RNG* rng, byte* priv, word32* privSz)
 {
     int ret;
     word32 sz = mp_unsigned_bin_size(&key->p);
-    sz = min(sz, 2 * DiscreteLogWorkFactor(sz * WOLFSSL_BIT_SIZE) /
+
+    /* predetermined values that operation would return for size */
+    WOLFSSL_DH_ROUND(sz); /* if using fixed points only, than round up */
+    switch (sz) {
```

Comment:
I think a comment noting where the sizes for the switch statement came from is in order, e.g., "Pre computed using the DiscreteLogWorkFactor algorithm and checked against the sizes in RFC 3526".